src/net.cpp

```diff
@@ -1239,20 +1239,16 @@ void CConnman::ThreadSocketHandler()
                             RecordBytesRecv(nBytes);
                             if (notify) {
                                 size_t nSizeAdded = 0;
+                                std::list<CNetMessage> completeMessages;
                                 auto it(pnode->vRecvMsg.begin());
                                 for (; it != pnode->vRecvMsg.end(); ++it) {
                                     if (!it->complete())
                                         break;
                                     nSizeAdded += it->vRecv.size() + CMessageHeader::HEADER_SIZE;
                                 }
-                                {
-                                    LOCK(pnode->cs_vProcessMsg);
-                                    pnode->vProcessMsg.splice(pnode->vProcessMsg.end(), pnode->vRecvMsg, pnode->vRecvMsg.begin(), it);
-                                    pnode->nProcessQueueSize += nSizeAdded;
-                                    if (pnode->nProcessQueueSize >= nReceiveFloodSize)
-                                        pnode->fPauseRecv = true;
-                                }
-                                WakeMessageHandler();
+                                completeMessages.splice(completeMessages.end(), pnode->vRecvMsg, pnode->vRecvMsg.begin(), it);
+                                if (!QueueReceivedMessages(pnode, std::move(completeMessages), nSizeAdded))
+                                    pnode->fPauseRecv = true;
```

Comment:
By moving the setting of fPauseRecv outside of the cs_vProcessMsg lock you've introduced a race where the message processing thread may have already taken the next message and set fPauseRecv to false before this line (seems pretty unlikely, but its not unheard of). I'd suggest simply dropping this commit.